pkg/controller/machine/machine_controller.go

```diff
@@ -310,3 +324,92 @@ func (r *ReconcileMachine) getMachinesInCluster(ctx context.Context, namespace,
 
 	return machines, nil
 }
+
+// Writer implements io.Writer interface as a pass-through for klog.
+type Writer struct {
+	logFunc func(args ...interface{})
+}
+
+// Write passes string(p) into Writer's logFunc and always returns len(p)
+func (w Writer) Write(p []byte) (n int, err error) {
+	w.logFunc(string(p))
+	return len(p), nil
+}
+
+// onPodDeletedOrEvicted is called by drain.Helper, when the pod has been deleted or evicted
+func onPodDeletedOrEvicted(pod *corev1.Pod, usingEviction bool) {
+	var verbStr string
+	if usingEviction {
+		verbStr = "evicted"
+	} else {
+		verbStr = "deleted"
+	}
+	klog.Infof("pod %s/%s %s\n", pod.Namespace, pod.Name, verbStr)
+}
+
+func (r *ReconcileMachine) drainNode(cluster *clusterv1.Cluster, machine *clusterv1.Machine) error {
+	var kubeClient kubernetes.Interface
+	if cluster == nil {
+		var err error
+		kubeClient, err = kubernetes.NewForConfig(r.config)
+		if err != nil {
+			return fmt.Errorf("unable to build kube client: %v", err)
+		}
+	} else {
+		// Otherwise, proceed to get the remote cluster client and get the Node.
+		remoteClient, err := remote.NewClusterClient(r.Client, cluster)
+		if err != nil {
+			klog.Errorf("Error creating a remote client for cluster %q while deleting Machine %q, won't retry: %v",
+				cluster.Name, machine.Name, err)
+			return nil
+		}
+		var err2 error
+		kubeClient, err2 = kubernetes.NewForConfig(remoteClient.RESTConfig())
+		if err2 != nil {
+			klog.Errorf("Error creating a remote client for cluster %q while deleting Machine %q, won't retry: %v",
+				cluster.Name, machine.Name, err)
+			return nil
+		}
+	}
+
+	node, err := kubeClient.CoreV1().Nodes().Get(machine.Status.NodeRef.Name, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// If an admin deletes the node directly, we'll end up here.
+			klog.Infof("Could not find node from noderef, it may have already been deleted: %v", machine.Status.NodeRef.Name)
+			return nil
+		}
+		return fmt.Errorf("unable to get node %q: %v", machine.Status.NodeRef.Name, err)
+	}
+
+	drainer := &kubedrain.Helper{
+		Client:              kubeClient,
```

Comment:
Can we try to scope down this client to only the needed used interfaces?

Reply:
@vincepri I'm not sure what you mean here.

Reply:
Does this library need to use the whole kubernetes interface or can we use something like CoreV1? Or a combination of multiple interfaces?

Reply:
https://github.com/kubernetes-sigs/cluster-api/pull/1096/files#diff-3645eb67eea12248070220dabd0e5da8R103

That is expecting a full client, it looks like.  I don't think the V1 client has a V1() method, so it would break.  This goes back to not changing the drain library.  In any case, the kubectl drain library uses the apigroup-specific clients internally.